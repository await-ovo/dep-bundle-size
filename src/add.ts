import { spawn } from 'node:child_process';
import { spinner } from '@clack/prompts';
import { bold } from 'kolorist';
import {
  detectPackageManager,
  findRootDirectory,
  formatSize,
  getConfig,
  logger,
  requestStats,
} from './utils';
import type { PackageStatsResponse, BundlephobiaConfig } from './utils';

const checkConstraints = ({
  constraints,
  config,
  stats,
  warning,
}: {
  constraints: { key: keyof BundlephobiaConfig; name: string }[];
  stats: PackageStatsResponse[];
  config: BundlephobiaConfig;
  warning?: boolean;
}) => {
  for (const constraint of constraints) {
    const limit = config[constraint.key];
    if (!limit) {
      continue;
    }
    const exceeded = stats.filter(({ size }) => size > limit);

    const hint = `Can not install ${exceeded
      .map(({ name, version, size }) =>
        bold(`${name}@${version}(${formatSize(size)})`),
      )
      .join(',')}, Since their ${
      constraint.name
    } is larger than the configured(${bold(formatSize(limit))}).`;
    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }

  const { overallSize, overallGzipSize } = stats.reduce<{
    overallSize: number;
    overallGzipSize: number;
  }>(
    (sizes, { size, gzip }) => {
      sizes.overallSize += size;
      sizes.overallGzipSize += gzip;
      return sizes;
    },
    {
      overallSize: 0,
      overallGzipSize: 0,
    },
  );

  if (config.maxOverallSize && overallSize > config.maxOverallSize) {
    const hint = `Can not install ${stats
      .map(({ name, version, size }) =>
        bold(`${name}@${version}(${formatSize(size)})`),
      )
      .join(',')}, Since their total size is larger than the configured(${bold(
      formatSize(config.maxOverallSize),
    )}).`;

    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }

  if (config.maxGzipSize && overallGzipSize > config.maxGzipSize) {
    const hint = `Can not install ${stats
      .map(
        ({ name, version, gzip }) =>
          `${name}@${version}(${bold(formatSize(gzip))})`,
      )
      .join(',')}, Since their total size is larger than the configured(${bold(
      formatSize(config.maxGzipSize),
    )}).`;
    if (warning) {
      logger.warn(hint);
    } else {
      throw new Error(hint);
    }
  }
};

const doAdd = async ({
  packages,
  cwd,
  rootProjectDir,
  dryRun,
}: {
  packages: string[];
  cwd: string;
  rootProjectDir: string;
  dryRun?: boolean;
}) => {
  const packageManager = detectPackageManager(rootProjectDir);

  const args = [
    packageManager,
    packageManager === 'npm' ? 'install' : 'add',
    ...packages,
  ];
  logger.log(`Add deps(${bold(args.join(' '))})`);
  if (!dryRun) {
    spawn(args[0], args.slice(1), {
      cwd,
      stdio: 'inherit',
    });
  }
};

export const add = async (
  packages: string[],
  options: {
    warning?: boolean;
    dir?: string;
    dryRun?: boolean;
  },
) => {
  const checkSpinner = spinner();
  checkSpinner.start('Checking constraints');
  const rootProjectDir = findRootDirectory(options.dir ?? process.cwd());
  const config = await getConfig(rootProjectDir);

  const stats = await requestStats(packages);

  checkConstraints({
    constraints: [
      { name: 'max-size', key: 'maxSize' },
      { name: 'max-gzip-size', key: 'maxGzipSize' },
    ],
    config,
    warning: options.warning,
    stats,
  });

  checkSpinner.stop('Checking constraints completed');

  await doAdd({
    packages,
    cwd: options.dir ?? process.cwd(),
    rootProjectDir,
    dryRun: options.dryRun,
  });
};
