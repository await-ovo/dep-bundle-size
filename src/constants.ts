export const BUNDLE_PHOBIA_API = 'https://bundlephobia.com/api/size';

export enum ErrorType {
  ParseJSONError,
}

export const ERROR_MESSAGES = {
  [ErrorType.ParseJSONError]: ({ file }: { file: string }) =>
    `JSON parse failed. file: ${file}`,
};
