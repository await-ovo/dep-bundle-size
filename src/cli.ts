#!/usr/bin/env node
import { cac } from 'cac';
import { scan } from '.';
import { add } from './add';
import { version } from '../package.json';

const cli = cac('dep-bundle-size');

cli.help();
cli.version(version);

cli
  .command('[...packages]')
  .option('-i, --interactive', 'Select packages to scan')
  .option('-r, --recursive', 'Select packages to scan')
  .action(scan);

cli
  .command('add <...packages>')
  .option('-i, --interactive', 'should report interactiveliy')
  .option('-w, --warning', 'use warning instead error')
  .action(add);

cli.parse();
