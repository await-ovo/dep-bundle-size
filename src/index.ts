import path from 'node:path';
import os from 'os';
import enquirer from 'enquirer';
import { table } from 'table';
import pLimit from 'p-limit';
import ora from 'ora';
import {
  findWorkspaceProjects,
  formatSize,
  formatTime,
  getDependencies,
  getDepInstalledVersion,
  getTimeFromSize,
  logger,
  requestStats,
  RuntimeError,
} from './utils';
import { ErrorType } from './constants';
import type { Project } from './utils';

const { prompt } = enquirer;

const doScan = async ({
  project,
  projects,
  rootProjectDir,
  packageArgs,
  interactive,
}: {
  project: Project;
  projects: Project[];
  rootProjectDir: string;
  packageArgs?: string[];
  interactive?: boolean;
}) => {
  const { path: projectDir } = project;
  const relativeProjectPath = path.relative(rootProjectDir, projectDir);
  const relativeName = relativeProjectPath === '' ? './' : relativeProjectPath;

  const spinner = ora({
    text: `Project ${relativeName} ...`,
    color: 'yellow',
  }).start();

  const deps = getDependencies(path.join(projectDir, 'package.json'), projects);

  let packages = packageArgs?.length ? packageArgs : Object.keys(deps);

  if (interactive) {
    ({ packages } = await prompt({
      type: 'multiselect',
      name: 'packages',
      message: 'Select packages to scan?',
      choices: packages,
    }));
  }

  if (!packages.length) {
    logger.warn(`No Packages to scan in project: ${projectDir}`);
    spinner.fail();
    return;
  }

  const stats = await requestStats(
    packages.map(
      packageName =>
        `${packageName}@${
          getDepInstalledVersion(packageName, projectDir) ??
          deps[packageName] ??
          'latest'
        }`,
    ),
  );

  spinner.text = `Project ${relativeName} completed:`;
  spinner.succeed();

  /* eslint-disable no-console */
  console.log(
    table(
      [
        ['Name', 'MIN', 'MIN + GZIP', 'SLOW 3G', 'EMERGING 4G'],
        ...stats.map(record => {
          const time = getTimeFromSize(record.gzip);
          return [
            `${record.name}@${record.version}`,
            formatSize(record.size),
            formatSize(record.gzip),
            formatTime(time.threeG),
            formatTime(time.fourG),
          ];
        }),
      ],
      {
        header: {
          alignment: 'center',
          content: 'Scan Results',
        },
      },
    ),
  );
  /* eslint-enable no-console */
};

export const scan = async (
  packageArgs: string[],
  options: { interactive?: boolean; recursive?: boolean },
) => {
  if (options.recursive && options.interactive) {
    throw new RuntimeError(ErrorType.RecursiveWithInteractiveError);
  }

  const rootProjectDir = process.cwd();

  const projects: Project[] = [
    {
      name: '.',
      path: rootProjectDir,
    },
  ];

  if (options.recursive) {
    projects.push(...(await findWorkspaceProjects(rootProjectDir)));
  }

  const limit = pLimit(os.cpus().length);

  await Promise.all(
    projects.map(project =>
      limit(() =>
        doScan({
          project,
          rootProjectDir,
          projects,
          packageArgs,
          interactive: options.interactive,
        }),
      ),
    ),
  );
};
