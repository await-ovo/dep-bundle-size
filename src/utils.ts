import fs from 'node:fs';
import path from 'node:path';
import { got, HTTPError } from 'got';
import pLimit from 'p-limit';
import yaml from 'js-yaml';
import glob from 'fast-glob';
import { BUNDLE_PHOBIA_API, ErrorType, NPM_REGISTRY } from './constants';

const limit = pLimit(6);

export type Project = {
  name?: string;
  path: string;
};

export type PackageStatsResponse = {
  assets: {
    gzip: number;
    name: string;
    size: number;
    type: string;
  }[];
  dependencyCount: number;
  dependencySizes: {
    approximateSize: number;
    name: string;
  }[];
  description: string;
  gzip: number;
  hasJSModule: string;
  hasJSNext: string;
  hasSideEffects: string[];
  name: string;
  isModuleType: boolean;
  repository: string;
  scoped: boolean;
  size: number;
  version: string;
};

export const logger = {
  log: (...args: any[]) => {
    process.stdout.write('\nINFO: ');
    // eslint-disable-next-line no-console
    console.log('[INFO]: ', ...args);
  },

  error: (...args: any[]) => {
    process.stderr.write('\nERROR: ');
    // eslint-disable-next-line no-console
    console.error(...args);
  },
  warn: (...args: any[]) => {
    process.stdout.write('\nWARN: ');
    // eslint-disable-next-line no-console
    console.warn(...args);
  },
};

export class RuntimeError extends Error {
  errorType: ErrorType;
  options?: any;
  constructor(errorType: ErrorType, options?: any) {
    super();
    this.errorType = errorType;
    this.options = options;
  }
}

export type PackageManager = 'pnpm' | 'yarn' | 'npm';

export const detectPackageManager = (dir: string) => {
  if (fs.existsSync(path.join(dir, 'pnpm-lock.yaml'))) {
    return 'pnpm';
  } else if (fs.existsSync(path.join(dir, 'package-lock.json'))) {
    return 'npm';
  } else if (fs.existsSync(path.join(dir, 'yarn.lock'))) {
    return 'yarn';
  }

  return 'npm';
};

export const getDepInstalledVersion = (
  specifier: string,
  dir?: string,
): string | undefined => {
  try {
    const packagePath = path.join(
      dir ?? process.cwd(),
      `node_modules/${specifier}/package.json`,
    );

    return JSON.parse(fs.readFileSync(packagePath, 'utf8')).version;
  } catch (err) {
    logger.warn(
      `Can not get installed version of package ${specifier}, use version listed in package.json or latest instead`,
    );
  }
};

export const getDependencies = (file: string, projects: Project[]) => {
  try {
    const json = JSON.parse(fs.readFileSync(file, 'utf8'));
    const dependencies = json.dependencies ?? {};

    // remove local dependencies
    return Object.keys(dependencies)
      .filter(depName => !projects.find(project => project.name === depName))
      .reduce<Record<string, string>>((deps, depName) => {
        deps[depName] = dependencies[depName];
        return deps;
      }, {});
  } catch (err) {
    throw new RuntimeError(ErrorType.ParseJSONError, { file });
  }
};

const memoize = <T>(fn: (...args: string[]) => Promise<T>) => {
  const cached = new Map<string, T>();

  return async (...args: string[]) => {
    const key = args.join('_');
    if (!cached.has(key)) {
      const res = await fn(...args);
      cached.set(key, res);
      return res;
    } else {
      return cached.get(key)!;
    }
  };
};

const requestPackageStats = memoize<PackageStatsResponse>(
  async (specifier: string) => {
    try {
      const packageStats: PackageStatsResponse = await got(
        `${BUNDLE_PHOBIA_API}?package=${specifier}&record=true`,
        {
          headers: {
            'User-Agent': 'bundle-phobia-cli',
            'X-Bundlephobia-User': 'bundle-phobia-cli',
          },
        },
      ).json();
      return packageStats;
    } catch (err) {
      if (err instanceof HTTPError) {
        const { code, message } = JSON.parse(err.response.body as string).error;
        throw new RuntimeError(ErrorType.RequestStatsError, {
          specifier,
          code,
          message,
        });
      } else {
        throw err;
      }
    }
  },
);

export const requestStats = async (specifiers: string[]) =>
  Promise.all(
    specifiers.map(specifier => limit(() => requestPackageStats(specifier))),
  );

export const findWorkspaceProjects = async (
  dir: string,
): Promise<Project[]> => {
  const packageManager = detectPackageManager(dir);

  let workspacePatterns: string[] = [];

  // pnpm workspaces
  if (
    packageManager === 'pnpm' &&
    fs.existsSync(path.join(dir, 'pnpm-workspace.yaml'))
  ) {
    const yamlContent = yaml.load(
      fs.readFileSync(path.join(dir, 'pnpm-workspace.yaml'), 'utf8'),
    ) as { packages?: string[] };
    workspacePatterns = yamlContent.packages ?? [];
  } else {
    const json = JSON.parse(
      fs.readFileSync(path.join(dir, 'package.json'), 'utf8'),
    );

    workspacePatterns = json.workspaces ?? [];
  }

  return workspacePatterns?.length
    ? (
        await glob(workspacePatterns, {
          onlyDirectories: true,
          unique: true,
          absolute: true,
        })
      )
        .filter(matched => fs.existsSync(path.join(matched, 'package.json')))
        .map(directory => ({
          name: (
            JSON.parse(
              fs.readFileSync(path.join(directory, 'package.json'), 'utf8'),
            ) as { name?: string }
          ).name,
          path: directory,
        }))
    : [];
};

// To be consistent with bundlephobia Web UI, the formatTime and getTimeFromSize utility functions modified from:
// https://github.com/pastelsky/bundlephobia/blob/bundlephobia/utils/index.js#L27
export const formatSize = (value: number) => {
  let unit, size;
  if (Math.log10(value) < 3) {
    unit = 'B';
    size = value;
  } else if (Math.log10(value) < 6) {
    unit = 'kB';
    size = value / 1024;
  } else {
    unit = 'MB';
    size = value / 1024 / 1024;
  }

  return `${size.toFixed(2)}${unit}`;
};

export const formatTime = (value: number) => {
  let unit, size;
  if (value < 0.0005) {
    unit = 'Î¼s';
    size = Math.round(value * 1000000);
  } else if (value < 0.5) {
    unit = 'ms';
    size = Math.round(value * 1000);
  } else {
    unit = 's';
    size = value;
  }

  return `${size}${unit}`;
};

export const getTimeFromSize = (sizeInBytes: number) => {
  return {
    threeG: sizeInBytes / 1024 / (400 / 8),
    fourG: sizeInBytes / 1024 / (7000 / 8),
  };
};
