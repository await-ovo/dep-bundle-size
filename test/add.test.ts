import path from 'node:path';
import { expect, test } from 'vitest';
import { createConsoleMock } from './utils';
import { add } from '../src/add';

const fixtures = path.join(process.cwd(), 'test/fixtures');

test('should add dependency in monorepo package successfully', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await add(['styled-components@5.3.6 '], {
    dir: path.join(fixtures, 'pnpm-workspace/bar'),
    warning: true,
    dryRun: true,
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});

test('should throw an error about exceeding size config', async () => {
  try {
    await add(['styled-components@5.3.6 '], {
      dir: path.join(fixtures, 'base-project'),
    });
  } catch (err: any) {
    expect(err?.message).toContain('Can not install styled-components@5.3.6');
  }
});

test('should ignore exceeding size error', async () => {
  const { getOutput, startMock, stopMock } = createConsoleMock();
  startMock();
  await add(['styled-components@5.3.6 '], {
    dir: path.join(fixtures, 'base-project'),
    warning: true,
    dryRun: true,
  });
  stopMock();
  expect(getOutput()).toMatchSnapshot();
});
