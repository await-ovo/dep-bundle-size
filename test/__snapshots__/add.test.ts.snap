// Vitest Snapshot v1

exports[`should add dependency in monorepo package successfully 1`] = `
"Info:  Add deps(pnpm add styled-components@5.3.6 )
"
`;

exports[`should ignore exceeding size error 1`] = `
"Warn:  Can not install styled-components@5.3.6(33.39kB), Since their max-size is larger than the configured(10.00B).
Info:  Add deps(npm install styled-components@5.3.6 )
"
`;
